### PR TITLE
AP_RangeFinder: add LeddarOne support

### PR DESCRIPTION
**Pull request by ShingoMatsuura**
add LeddarOne support

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_LeddarOne.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <ctype.h>
+
+extern const AP_HAL::HAL& hal;
+
+// Table for CRC values high byte
+static uint8_t CRC_HI[] =
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Const?

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        }
+    }
+
+    // parse a response message, set detections and sum_distance
+    if (parse_response() < 0) {
+        return false;
+    }
+
+    // calculate average distance
+    reading_cm = sum_distance / number_detections;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Divide by zero check needed

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    while (!uart->available())
+    {
+        // wait up to 200ms
+        if (AP_HAL::millis() - start_ms > 200) {
+            return false;
+        }
+    }
+
+    // parse a response message, set detections and sum_distance
+    if (parse_response() < 0) {
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: When num detected is zero, this returns zero and will cause divide by zero below

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 1 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Thank you for reviewing. I'm going to modify.

**Comment by muramura** [comment]
I think so.
We loop in AP_RangeFinder_LeddarOne :: parse_response method.
Loop times out.
The value of len is 0, and time-out.
The value obtained by subtracting 2 from the value of len sets the CRC16 method.
In this case, the underflow occurs.

This does not occur?

If occur, check the value of len.

**Comment by magicrub** [comment]
Why is the whole CRC hi/low stored? Can it be computed easily on demand? Seems like a waste of 512 bytes of program memory.

**Comment by ShingoMatsuura** [comment]
Thank you for comment.

I am going to check if the len value is 25.If it is not 25, the response is bad.

Regarding to CRC, I saw leddar tech sdk sample code. It uses CRC hi/low tables. I am going to try to change the CRC function to on demand.

**Comment by magicrub** [comment]
@ShingoMatsuura it may be required, CRC's are tricky.

**Comment by rmackay9** [comment]
That is great that you managed to get rid of the CRC table!

**Comment by muramura** [comment]
I think.
The information in the uart has acquired in **AP_SerialManager :: SerialProtocol_Lidar**.
Here, use the value you have added a definition of **leddarone** to **AP_SerialManager**.
This is because it is not affected by the change of the **lidar**.

**Comment by ShingoMatsuura** [comment]
Thank you for your comment. 

The definition of RangeFinder_TYPE_LEDDARONE is used in RangeFinder::detect_instance. It is read from RNGFND_TYPE parameter. If you want to use LeddarOne, you need to set RNGFND_TYPE = 12.

For AP_SerialManager, yon need to set as below.

SERIAL4_PROTOCOL = 9 (Lidar)
SERIAL4_BAUD = 115

**Comment by muramura** [comment]
The AP_Common.h have been defined LOWBYTE and HIGHBYTE of uint16_t.
What about to use here.

**Comment by ShingoMatsuura** [comment]
Thank your for your advice.

I think It is better. I am going to change CRC function and test.

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by rmackay9** [comment]
I've created a wiki page for it too: http://ardupilot.org/copter/docs/common-leddar-one-lidar.html

**Pull request closed without merging**